tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ManagedIdentityTests.cs

```diff
@@ -480,5 +483,168 @@ public async Task ManagedIdentityCacheTestAsync()
+        public async Task ManagedIdentityIsProactivelyRefreshedAsync()
+        {
+            using (new EnvVariableContext())
+            using (var httpManager = new MockHttpManager())
+            {
+                SetEnvironmentVariables(ManagedIdentitySource.AppService, AppServiceEndpoint);
+
+                Trace.WriteLine("1. Setup an app with a token cache with one AT");
+
+                var mi = ManagedIdentityApplicationBuilder.Create()
+                    .WithExperimentalFeatures()
+                    .WithHttpManager(httpManager)
+                    .BuildConcrete();
+
+                httpManager.AddManagedIdentityMockHandler(
+                        AppServiceEndpoint,
+                        Resource,
+                        MockHelpers.GetMsiSuccessfulResponse(),
+                        ManagedIdentitySource.AppService);
+
+                AuthenticationResult result = await mi.AcquireTokenForManagedIdentity(Resource)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                Trace.WriteLine("2. Configure AT so that it shows it needs to be refreshed");
+                var refreshOn = UpdateATWithRefreshOn(mi.AppTokenCacheInternal.Accessor).RefreshOn;
+                TokenCacheAccessRecorder cacheAccess = mi.AppTokenCache.RecordAccess();
+
+                Trace.WriteLine("3. Configure MSI to respond with a valid token");
+                httpManager.AddManagedIdentityMockHandler(
+                        AppServiceEndpoint,
+                        Resource,
+                        MockHelpers.GetMsiSuccessfulResponse(),
+                        ManagedIdentitySource.AppService);
+
+                // Act
+                Trace.WriteLine("4. ATM - should perform an RT refresh");
+                result = await mi.AcquireTokenForManagedIdentity(Resource)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+
+                // Assert
+                YieldTillSatisfied(() => httpManager.QueueSize == 0);
+
+                Assert.IsNotNull(result);
+
+                Assert.AreEqual(0, httpManager.QueueSize,
+                    "MSAL should have refreshed the token because the original AT was marked for refresh");
+                
+                cacheAccess.WaitTo_AssertAcessCounts(1, 1);
+
+                Assert.IsTrue(result.AuthenticationResultMetadata.CacheRefreshReason == CacheRefreshReason.ProactivelyRefreshed);
+                
+                Assert.IsTrue(result.AuthenticationResultMetadata.RefreshOn == refreshOn);
+
+                result = await mi.AcquireTokenForManagedIdentity(Resource)
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+                
+                Assert.IsTrue(result.AuthenticationResultMetadata.CacheRefreshReason == CacheRefreshReason.NotApplicable);
+            }
+        }
+
+        private bool YieldTillSatisfied(Func<bool> func, int maxTimeInMilliSec = 30000)
```

Comment:
This is used in multiple tests. Should be extracted in some common class.

Ideally, long term, we should refactor to not need this. Very flaky. :(